### DISCUSSION TITLE
高可用分布式存储 etcd 的实现原理 · /etcd-introduction

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#531] [Announcements]
https://draveness.me/etcd-introduction

**Comment by wujunze** [comment]
干货  👍

**Comment by lcullen** [comment]
想请教一下作者，当一个节点发起投票msg的时候，其他节点收到消息，拒绝投票的原因是不是:当前节点自己的事务id 要大于 请求过来的 事务id

**Comment by draveness** [comment]
> @lcullen
> 想请教一下作者，当一个节点发起投票msg的时候，其他节点收到消息，拒绝投票的原因是不是:当前节点自己的事务id 要大于 请求过来的 事务id

这应该只是其中一个拒绝投票的原因

**Comment by Triple-Z** [comment]
听了作者做的那次播客找到了这里，博文质量真的很高，向作者学习！

**Comment by solozyx** [comment]
> @Triple-Z
> 听了作者做的那次播客找到了这里，博文质量真的很高，向作者学习！

什么播客 能分享一下 播客的链接 么

**Comment by alegjk** [comment]
谢谢分享，内容很不错。 请教下， 文章中的图用什么工具画的啊？

**Comment by wujunze** [comment]
> > @Triple-Z
> > 听了作者做的那次播客找到了这里，博文质量真的很高，向作者学习！
> 
> 什么播客 能分享一下 播客的链接 么

 哪个播客  能不能发个链接

**Comment by Triple-Z** [comment]
> 
> 
> > > @Triple-Z
> > > 听了作者做的那次播客找到了这里，博文质量真的很高，向作者学习！
> > 
> > 
> > 什么播客 能分享一下 播客的链接 么
> 
> 哪个播客 能不能发个链接

@wujunze @solozyx 
https://talk.swift.gg/1

**Comment by kelvinkuo** [comment]
想知道用的什么绘图软件

**Comment by draveness** [comment]
> @kelvinkuo
> 想知道用的什么绘图软件

https://draveness.me/sketch-and-sketch

**Comment by chenjinxuan** [comment]
如果当前节点可以成为 Leader 并且上一次收到 Leader 节点的消息或者心跳已经超过了等待的时间，当前节点就会发送 MsgHup 消息尝试开始新的选举。

上面这句话里说 ”如果当前节点可以成为 Leader“ 是根据什么机制来判断当前节点可以成为Leader的呢

**Comment by draveness** [comment]
> 如果当前节点可以成为 Leader 并且上一次收到 Leader 节点的消息或者心跳已经超过了等待的时间，当前节点就会发送 MsgHup 消息尝试开始新的选举。
> 
> 上面这句话里说 ”如果当前节点可以成为 Leader“ 是根据什么机制来判断当前节点可以成为Leader的呢

超时和选举投票

**Comment by xiaoheiAh** [comment]
想问大佬一个raft的场景：
leader 收到客户端请求后存到本地log，再通过 rpc 将log追加到其他的节点上，发出请求后，leader down掉了，假设超过半数的follower收到了该log，并存了下来，只是没有commit，在剩下的节点中选举出一个 leader 后，leader 收到一个新请求进来，就会将之前的这个未 commit 的 log 顺带 commit 掉。这样不就出现问题了吗？之前的那个请求，客户端会以为失败了，但实际在后面的term中他还是被提交掉了

**Comment by draveness** [comment]
> 想问大佬一个raft的场景：
> leader 收到一个新请求进来，就会将之前的这个未 commit 的 log 顺带 commit 掉。

这个是啥意思，什么叫『 commit 的 log 顺带 commit 掉』

**Comment by xiaoheiAh** [comment]
> > 想问大佬一个raft的场景：
> > leader 收到一个新请求进来，就会将之前的这个未 commit 的 log 顺带 commit 掉。
> 
> 这个是啥意思，什么叫『 commit 的 log 顺带 commit 掉』

就是leader在commit这条日志前down掉了,但是这条日志已经在其他的节点上了,都还是uncommitted的状态,下一个leader选举出来后,如果追加日志就会把这条日志页提交掉.就是论文里 5.4.2 Committing entries from previous terms 说的逻辑

**Comment by draveness** [comment]
> > > 想问大佬一个raft的场景：
> > > leader 收到一个新请求进来，就会将之前的这个未 commit 的 log 顺带 commit 掉。
> > 
> > 
> > 这个是啥意思，什么叫『 commit 的 log 顺带 commit 掉』
> 
> 就是leader在commit这条日志前down掉了,但是这条日志已经在其他的节点上了,都还是uncommitted的状态,下一个leader选举出来后,如果追加日志就会把这条日志页提交掉

你确定是提交不是 rollback 么，或者说覆盖

**Comment by xiaoheiAh** [comment]
![raft](https://user-images.githubusercontent.com/27981726/89028986-fab8f280-d35f-11ea-9f49-11ea9a949906.gif)

应该能描述出这个场景了

**Comment by draveness** [comment]
> 
> 
> ![raft](https://user-images.githubusercontent.com/27981726/89028986-fab8f280-d35f-11ea-9f49-11ea9a949906.gif)
> 
> 
> 
> 应该能描述出这个场景了
> 
> 

你这个图演示的非常好，我想了一下遇到这种问题客户端可能会收到超时或者未知错误，它需要重试获得确定的结果，而接口需要通过唯一 ID 保证幂等

**Comment by xiaoheiAh** [comment]
> > ![raft](https://user-images.githubusercontent.com/27981726/89028986-fab8f280-d35f-11ea-9f49-11ea9a949906.gif)
> > 应该能描述出这个场景了
> 
> 你这个图演示的非常好，我想了一下遇到这种问题客户端可能会收到超时或者未知错误，它需要重试获得确定的结果，而接口需要通过唯一 ID 保证幂等

应该也只有通过这种方式去处理了，貌似无法避免这种情况

**Comment by Mrdshu** [comment]
> @xiaoheiAh
> 
> ![raft](https://user-images.githubusercontent.com/27981726/89028986-fab8f280-d35f-11ea-9f49-11ea9a949906.gif)
> 
> 应该能描述出这个场景了
> 

请问这个图有链接吗，想去操作一把

**Comment by Triple-Z** [comment]
@Mrdshu 应该是这里吧：https://raft.github.io/

**Comment by liuyu121** [comment]
图文并茂，真心牛逼

**Comment by Ethan-rookie** [comment]
咨询下大佬这个任期是怎么配置的呢

**Comment by draveness** [comment]
> 咨询下大佬这个任期是怎么配置的呢

只有主节点超时之后才会重新选主，相关的配置看看这个 https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/configuration.md#--election-timeout

**Comment by Janetyu** [comment]
每当一个 Key 被删除时都会调用 timestone 方法向当前的 generation 中追加一个新的墓碑版本

timestone -》 tombstone？

---

*2020-10-11 UPDATES: 已修复*

**Comment by Ethan-rookie** [comment]
> @draveness
> > 咨询下大佬这个任期是怎么配置的呢
> 
> 只有主节点超时之后才会重新选主，相关的配置看看这个 https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/configuration.md#--election-timeout

多谢，想再咨询个问题，我看到k8s里面取mvcc中keyvalue结构体中的ModRevision字段作为资源的resourceversion,来源于事务的id，同个事务可能操作多个对象，这样最终k8s里面两个对象的resourcevision不就一样了吗，怎么去区分呢，是不是我理解的有问题

**Comment by draveness** [comment]
> > @draveness
> > > 咨询下大佬这个任期是怎么配置的呢
> > 
> > 
> > 只有主节点超时之后才会重新选主，相关的配置看看这个 https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/configuration.md#--election-timeout
> 
> 多谢，想再咨询个问题，我看到k8s里面取mvcc中keyvalue结构体中的ModRevision字段作为资源的resourceversion,来源于事务的id，同个事务可能操作多个对象，这样最终k8s里面两个对象的resourcevision不就一样了吗，怎么去区分呢，是不是我理解的有问题

Kubernetes 可以同时操作多个 Resource 么？

**Comment by wujunwei** [comment]
> 对于 kvstore 来说，其实没有太多值得展开介绍的地方，它利用底层的 BoltDB 等基础设施为上层提供最常见的增伤改查

增伤改查 => 增删改查
